Add rechunking for Xarray datasets

eric-czech opened the pull request:
This is an attempt at https://github.com/pangeo-data/rechunker/issues/45.

I'm not sure what the best way to go about this is, but I thought I would get something working and then get thoughts from you guys on where to go next.  Notes:

- The signature I'm exposing here is `rechunk_dataset(source: Dataset, encoding: Mapping, max_mem, target_store, temp_store, executor)`.  I'm using `encoding` to indicate the target chunkings, along with any other compressor/filter options, so there is some consistency with `Dataset.to_zarr`.  I think it would probably be best if `Dataset`/`DataArray` were other possible options in the main `rechunk` function with the same `target_chunks` parameter and any other options in `{target|temp}_options`.  For the sake of discussion I thought it was easier to review quickly if all the new code was in one place.  I'll happily combine the functions if this is on the right track.
- I'm probably missing some corner cases, but this seemed like a reasonable lift of the necessary functionality in the [zarr backend](https://github.com/pydata/xarray/blob/9ee0f018aac45a83df7a65b1499263412dab9bed/xarray/backends/zarr.py).  I tried to ensure that the condition validated at [xarray/backends/zarr.py#L135](https://github.com/pydata/xarray/blob/9ee0f018aac45a83df7a65b1499263412dab9bed/xarray/backends/zarr.py#L135) is also checked here.  I'm not actually sure if that matters in this case.  @rabernat / @TomAugspurger do you happen to know if I should need to worry about this? I.e. if the input dataset contains dask arrays backed by a zarr store and the target chunking would create new dask chunks that overlap multiple old zarr chunks, should rechunker throw an error like xarray does?
- I'm encoding attributes using [encode_zarr_attr_value](https://github.com/pydata/xarray/blob/9ee0f018aac45a83df7a65b1499263412dab9bed/xarray/backends/zarr.py#L16).  Should all of the rechunker functions be using this or something like it too?
- I think there are several more tests that would be useful -- the test in this PR is just a start.

tomwhite commented:
This looks great @eric-czech, thanks for working on it.

> I think it would probably be best if `Dataset`/`DataArray` were other possible options in the main `rechunk` function

+1

shoyer commented on rechunker/api.py:
```diff
@@ -177,6 +185,70 @@ def _get_executor(name: str) -> Executor:
+
+    copy_specs = []
+    for variable in source:
+        array = source[variable].copy()
```
The other part that is missing here is that you are only writing the data variables, not the coordinates. Iterating over `source.variables` should solve that problem, too.

eric-czech replied:
Ah got it -- made some associated changes starting at https://github.com/pangeo-data/rechunker/pull/52/files#diff-5f579dcc7633808e087447d18c990e51R315.  

Test with coordinates is at https://github.com/pangeo-data/rechunker/pull/52/files#diff-cbfcc614e55a07a4d1df274c34ed50b4R51.

shoyer commented:
> * I'm encoding attributes using [encode_zarr_attr_value](https://github.com/pydata/xarray/blob/9ee0f018aac45a83df7a65b1499263412dab9bed/xarray/backends/zarr.py#L16).  Should all of the rechunker functions be using this or something like it too?

I suspect this is only really relevant if the source is from xarray:

- If the source is from zarr, attributes are already encoded properly
- If the source is from dask, attributes don't exist

eric-czech commented:
Ok I re-worked this one a good bit (apologies for the big changes since first review).  Some notes on the latest commit at https://github.com/pangeo-data/rechunker/pull/52/commits/fc1b17a6629644ff678f3655dddfe65397b80fcc:

- Added Datasets as a possible source in `api.rechunk` (there is no `rechunk_dataset` now)
- Changed `{temp|target}_options` to be specific to individual variables and have a structure similar to `target_chunks`
- Defaulted the temporary store location to somewhere in the system temp dir for Zarr groups and Xarray datasets
  - It didn't make much sense IMO for `temp_store` to be optional in the API and not have a default be set for collections of arrays.  Otherwise when it's not set, an assertion error was thrown if any one array is rechunked to a different size.
- I skipped trying to support DataArrays because of https://github.com/pydata/xarray/issues/2660
- The happy path for these changes would result from simple usage like this:

```python
import zarr
import xarray as xr
import numpy as np
from rechunker.api import rechunk

shape = (100, 50)
ds = xr.Dataset(
    dict(
        a=(("x", "y"), np.ones(shape, dtype='f4')),
        b=(("x"), np.ones(shape[0])),
        c=(("y"), np.ones(shape[1]))
    ),
    coords=dict(
        cx=(("x"), np.ones(shape[0])),
        cy=(("y"), np.ones(shape[1]))
    )
).chunk(chunks=25)

rechunked = rechunk(
    ds,
    target_chunks=dict(a=(10, 10), b=(10,), c=(10,)),
    max_mem='50MB',
    target_store="/tmp/store5.zarr",
    target_options=dict(
        a=dict(
            compressor=zarr.Blosc(cname="zstd"),
            dtype="int16",
            scale_factor=0.1,
            _FillValue=-9999,
        )
    )
)
```

shoyer reviewed (commented):
> * Defaulted the temporary store location to somewhere in the system temp dir for Zarr groups and Xarray datasets
>   
>   * It didn't make much sense IMO for `temp_store` to be optional in the API and not have a default be set for collections of arrays.  Otherwise when it's not set, an assertion error was thrown if any one array is rechunked to a different size.

I would rather require an explicit temp directory for now. My concern is that using a local directory as a default is likely to result in unexpected errors when scaling up rechunker for "production" use cases that run on multiple machines. Perhaps there is some way we might ask Executors to provide a location for temporary storage

Note that in the future there will be Executors that don't require temporary arrays on disk (e.g., see Beam in https://github.com/pangeo-data/rechunker/pull/36).

rabernat commented:
Thanks for all the hard work happening here!

> I would rather require an explicit temp directory for now. My concern is that using a local directory as a default is likely to result in unexpected errors when scaling up rechunker for "production" use cases that run on multiple machines.

👍  to this. Our main use of rechunker is using dask in the cloud with object store, where there is no shared local filesystem. I'd like to avoid any default assumptions about the nature of the storage.

Going forward, maybe we could consider adding some sort of config system for rechunker, which would allow you to specify your preferred way of creating temporary storage.

eric-czech commented on tests/test_rechunk.py:
```diff
@@ -35,6 +37,75 @@ def target_chunks(request):
+@pytest.mark.parametrize("target_chunks", [(20, 10)])
+@pytest.mark.parametrize("max_mem", ["10MB"])
+@pytest.mark.parametrize("pass_temp", [True, False])
+@pytest.mark.parametrize("executor", ["dask", api._get_executor("dask")])
```
Hm it doesn't, but I can't tell if it should.  I'm converting all the arrays in the dataset to dask and sending them through the dask.array.Array codepath, which is apparently also broken for all other executors.  This is the trace I get in one form or another from different executors:

```
Traceback (most recent call last):
  File "/Users/eczech/repos/pydata/rechunker/tests/test_rechunk.py", line 100, in test_rechunk_dataset
    rechunked.execute()
  File "/Users/eczech/repos/pydata/rechunker/rechunker/api.py", line 77, in execute
    self._executor.execute_plan(self._plan, **kwargs)
  File "/Users/eczech/repos/pydata/rechunker/rechunker/executors/python.py", line 31, in execute_plan
    plan()
  File "/Users/eczech/repos/pydata/rechunker/rechunker/executors/python.py", line 44, in _execute_all
    task()
  File "/Users/eczech/repos/pydata/rechunker/rechunker/executors/python.py", line 39, in _direct_array_copy
    target[key] = source[key]
  File "/Users/eczech/.conda/envs/rechunker-dev/lib/python3.7/site-packages/zarr/core.py", line 1115, in __setitem__
    self.set_basic_selection(selection, value, fields=fields)
  File "/Users/eczech/.conda/envs/rechunker-dev/lib/python3.7/site-packages/zarr/core.py", line 1210, in set_basic_selection
    return self._set_basic_selection_nd(selection, value, fields=fields)
  File "/Users/eczech/.conda/envs/rechunker-dev/lib/python3.7/site-packages/zarr/core.py", line 1501, in _set_basic_selection_nd
    self._set_selection(indexer, value, fields=fields)
  File "/Users/eczech/.conda/envs/rechunker-dev/lib/python3.7/site-packages/zarr/core.py", line 1550, in _set_selection
    self._chunk_setitem(chunk_coords, chunk_selection, chunk_value, fields=fields)
  File "/Users/eczech/.conda/envs/rechunker-dev/lib/python3.7/site-packages/zarr/core.py", line 1665, in _chunk_setitem
    fields=fields)
  File "/Users/eczech/.conda/envs/rechunker-dev/lib/python3.7/site-packages/zarr/core.py", line 1687, in _chunk_setitem_nosync
    chunk = value.astype(self._dtype, order=self._order, copy=False)
  File "/Users/eczech/.conda/envs/rechunker-dev/lib/python3.7/site-packages/dask/array/core.py", line 1843, in astype
    "arguments: {0!s}".format(list(extra))
TypeError: astype does not take the following keyword arguments: ['order']
```

Two potential solutions I see are:

1. Coerce all dataset variables to zarr when not using the dask executor
2. Make the other executors work on dask arrays

The first seems like a bad idea, and I'm not sure how to do the second.

shoyer replied:
I suspect it would just to change `target[key] = source[key]` to `target[key] = np.asarray(source[key])`  inside `_direct_array_copy`? Or perhaps this fix could be done upstream inside `__setitem__` on Zarr arrays?

Anyways, this can definitely be saved for later!

eric-czech commented:
> Our main use of rechunker is using dask in the cloud with object store, where there is no shared local filesystem

Ah of course, makes sense.

In https://github.com/pangeo-data/rechunker/pull/52/commits/67ee2aa4674d94cbb11f18c04326504872745283, I removed the default temp store, added a better error when it's not present, and added a NotImplementedError when the source is Xarray and the executor is anything but dask.  I think I should probably do the same for when the source is da.Array.  Does this sound right to you both?

shoyer commented:
> In [67ee2aa](https://github.com/pangeo-data/rechunker/commit/67ee2aa4674d94cbb11f18c04326504872745283), I removed the default temp store, added a better error when it's not present, and added a NotImplementedError when the source is Xarray and the executor is anything but dask. I think I should probably do the same for when the source is da.Array. Does this sound right to you both?

This sounds fine to me for now.

Long term, I do think it could make sense to pass an `xarray.Dataset` backed by multi-threaded dask into alternative executors, such as Beam. But this certainly isn't urgent.

eric-czech commented:
> In 67ee2aa, I removed the default temp store, added a better error when it's not present, and added a NotImplementedError when the source is Xarray and the executor is anything but dask. I think I should probably do the same for when the source is da.Array. Does this sound right to you both?
>> This sounds fine to me for now.

Ok, https://github.com/pangeo-data/rechunker/pull/52/commits/8502a33df6b5c5d5416b9b81ed268a2daa4c75e8 adds a similar error for dask array sources.

> Long term, I do think it could make sense to pass an xarray.Dataset backed by multi-threaded dask into alternative executors, such as Beam. But this certainly isn't urgent.

I see, maybe the error in https://github.com/pangeo-data/rechunker/pull/52#discussion_r498457119 is actually pretty superficial?  I can't tell whether or not that's hinting at a fundamental limitation.

eric-czech commented:
Is there anything else you guys think I should address on this one?

eric-czech commented:
Hey @shoyer sorry to keep bugging you about this one, but is there anything else you'd like me to change?

rabernat commented:
Hi @eric-czech. Thanks for your work on this! And thanks for your patience.

I'm fine with merging now. I assume issues will come up as people try it out, and we can iterate as needed.

eric-czech commented:
Thanks @rabernat!  And for your suggestions @shoyer.
